File: app/src/main/java/com/funckyhacker/fileexplorer/view/search/SearchViewModelImpl.java
package com.funckyhacker.fileexplorer.view.search;

import android.os.Environment;
import android.text.TextUtils;
import com.funckyhacker.fileexplorer.view.adapter.MainLinearAdapter;
import java.io.File;
import java.util.ArrayList;
import java.util.List;
import javax.inject.Inject;
import rx.Emitter;
import rx.Observable;
import timber.log.Timber;

public class SearchViewModelImpl extends SearchViewModel {

  private MainLinearAdapter linearAdapter;

  @Inject
  public SearchViewModelImpl() {}

  @Override public Observable<List<File>> search(String query) {
    return Observable.create(e -> {
      long start = System.currentTimeMillis();
      List<File> result = new ArrayList<>();

      File rootDir = new File(Environment.getExternalStorageDirectory().getPath());
      if (!rootDir.isDirectory() || rootDir.listFiles() == null || rootDir.listFiles().length <= 0) {
        e.onError(new Throwable("Something wrong your External Storage"));
      }
      searchDir(result, rootDir, query, e);
      e.onNext(result);
      long end = System.currentTimeMillis();
      Timber.d("Search Elapsed Time: %d", end - start);
    }, Emitter.BackpressureMode.DROP);
  }

  @Override public void init() {
    linearAdapter = new MainLinearAdapter();
  }

  @Override public MainLinearAdapter getLinearAdapter() {
    return linearAdapter;
  }

  @Override public void setData(List<File> files) {
    linearAdapter.setData(files);
  }

  private void searchDir(List<File> result, File root, String query, Emitter<List<File>> e) {
    for (File file : root.listFiles()) {
      if (TextUtils.isEmpty(file.getName())){
        continue;
      }
      if (file.getName().contains(query) || file.getName().toLowerCase().contains(query)) {
        result.add(file);
      }
      if (file.isDirectory()) {
        searchDir(result, file, query, e);//Notice: Recursive Call
      }
    }
  }
}

File: app/src/main/java/com/funckyhacker/capofiler/view/search/SearchViewModelImpl.kt
package com.funckyhacker.capofiler.view.search

import android.os.Environment
import android.text.TextUtils
import com.funckyhacker.capofiler.view.adapter.MainLinearAdapter
import rx.Emitter
import rx.Observable
import java.io.File
import java.util.*
import javax.inject.Inject

class SearchViewModelImpl @Inject
constructor() : SearchViewModel() {

    override lateinit var linearAdapter: MainLinearAdapter
        private set

    override fun search(query: String): Observable<List<File>> {
        return Observable.create({ e ->
            val result = ArrayList<File>()

            val rootDir = File(Environment.getExternalStorageDirectory().path)
            if (!rootDir.isDirectory || rootDir.listFiles() == null || rootDir.listFiles().isEmpty()) {
                e.onError(Throwable("Something wrong your External Storage"))
            }
            searchDir(result, rootDir, query)
            e.onNext(result)
        }, Emitter.BackpressureMode.DROP)
    }

    override fun init() {
        linearAdapter = MainLinearAdapter()
    }

    override fun setData(files: List<File>) {
        linearAdapter.setData(files)
    }

    private fun searchDir(result: MutableList<File>, root: File, query: String) {
        for (file in root.listFiles()) {
            if (TextUtils.isEmpty(file.name)) {
                continue
            }
            if (file.name.contains(query) || file.name.toLowerCase().contains(query)) {
                result.add(file)
            }
            if (file.isDirectory) {
                searchDir(result, file, query)//Notice: Recursive Call
            }
        }
    }
}
